packages/zql/src/query/ast-to-zql.test.ts

```diff
@@ -0,0 +1,518 @@
+  const ast: AST = {
+    table: 'issue',
+    where: {
+      type: 'and',
+      conditions: [
+        {
+          type: 'simple',
+          left: {type: 'column', name: 'status'},
+          op: '!=',
+          right: {type: 'literal', value: 'closed'},
+        },
+        {
+          type: 'simple',
+          left: {type: 'column', name: 'priority'},
+          op: '>=',
+          right: {type: 'literal', value: 3},
+        },
+      ],
+    },
+    orderBy: [['created_at', 'desc']],
+    limit: 20,
+    related: [
+      {
+        correlation: {
+          parentField: ['id'],
+          childField: ['issue_id'],
+        },
+        subquery: {
+          table: 'comment',
+          alias: 'comments',
+          limit: 5,
+          orderBy: [['created_at', 'desc']],
+        },
+      },
+    ],
+  };
+  expect(astToZQL(ast)).toMatchInlineSnapshot(
+    `".where('status', '!=', 'closed').where('priority', '>=', 3).related('comments', q => q.orderBy('created_at', 'desc').limit(5)).orderBy('created_at', 'desc').limit(20)"`,
+  );
+});
+
+test('with auth parameter', () => {
+  const ast: AST = {
+    table: 'issue',
+    where: {
+      type: 'simple',
+      left: {type: 'column', name: 'owner_id'},
+      op: '=',
+      right: {
+        type: 'static',
+        anchor: 'authData',
+        field: 'id',
+      },
+    },
+  };
+  expect(astToZQL(ast)).toMatchInlineSnapshot(
+    `".where('owner_id', authParam('id'))"`,
+  );
+});
+
+test('EXISTS with order', () => {
+  const ast: AST = {
+    table: 'users',
+    orderBy: [['id', 'asc']],
+    where: {
+      type: 'correlatedSubquery',
+      related: {
+        correlation: {parentField: ['recruiterID'], childField: ['id']},
+        subquery: {
+          table: 'users',
+          alias: 'zsubq_recruiter',
+          where: {
+            type: 'simple',
+            left: {type: 'column', name: 'y'},
+            op: '>',
+            right: {type: 'literal', value: 0},
+          },
+        },
+      },
+      op: 'EXISTS',
+    },
+  };
+
+  expect(astToZQL(ast)).toMatchInlineSnapshot(
+    `".whereExists('recruiter', q => q.where('y', '>', 0))).orderBy('id', 'asc')"`,
+  );
+});
+
+test('round trip', () => {
+  const randomizer = generateMersenne53Randomizer(42);
+  const rng = () => randomizer.next();
+  const faker = new Faker({
+    locale: en,
+    randomizer,
+  });
+
+  const codes: string[] = [];
+
+  for (let i = 0; i < 10; i++) {
+    const schema = generateSchema(rng, faker, 10);
+    const q = generateQuery(schema, {}, rng, faker);
+
+    const code = astToZQL(ast(q));
+    codes.push(code);
+
+    const q2 = new Function(
+      'staticQuery',
+      'schema',
+      'tableName',
+      `return staticQuery(schema, tableName)${code}`,
+    )(staticQuery, schema, ast(q).table);
+    expect(ast(q2)).toEqual(ast(q));
+  }
+
+  expect(codes).toMatchInlineSnapshot(`
+    [
+      ".where('nudge', 'IS NOT', false)",
```

Comment:
generateQuery doesn't generate very advanced queries but it is still really cool to able to do this.